[RFE] Expose WOLF_CRYPTO_CB_RSA_PAD as a configure knob

### Version

master

### Description

RFE: expose the C defines `WOLF_CRYPTO_CB_RSA_PAD`  as a autotools/configure or cmake knob.

HSMs today may disable RSA raw operations and users are recommended to use RSA with padding mechanisms instead of managing padding in software. `wc_pkcs11.c` has support for OAEP/PKCS15/PSS mechanisms based on `WOLF_CRYPTO_CB_RSA_PAD`.  This flag seems to be defined only with `WOLFSSL_RENESAS_TSIP`.

```
#if defined(WOLFSSL_RENESAS_TSIP)
    #define TSIP_TLS_HMAC_KEY_INDEX_WORDSIZE 64
    #define TSIP_TLS_MASTERSECRET_SIZE       80   /* 20 words */
    #define TSIP_TLS_ENCPUBKEY_SZ_BY_CERTVRFY 560 /* in byte  */

    #ifdef WOLF_CRYPTO_CB
        /* make sure RSA padding callbacks are enabled */
        #define WOLF_CRYPTO_CB_RSA_PAD
    #endif
#endif /* WOLFSSL_RENESAS_TSIP */
```

While this flag can be turned on with `EXTRA_CFLAGS` it seems more ergonomic to allow it to be turned on at configure time. Currently Debian does not have this flag turned on (given that it is quite obscure). This means that `wc_pkcs11.c` may not be using optimal mechanisms.

Took a crack at this. Extended the existing `--enable-cryptocbutils` list with `rsapad` rather than adding a new top-level flag, so there's one place for crypto-callback sub-features. Added a matching `WOLFSSL_CRYPTOCB_RSA_PAD` option on the CMake side.

Usage:

```sh
# autotools — just the padding callbacks
./configure --enable-cryptocb --enable-cryptocbutils=rsapad

# combined with other utilities
./configure --enable-cryptocb --enable-cryptocbutils=copy,free,rsapad

# all utilities (now includes rsapad)
./configure --enable-cryptocb --enable-cryptocbutils

# pkcs11 forces cryptocb on, so this works too
./configure --enable-pkcs11 --enable-cryptocbutils=rsapad

# cmake
cmake -DWOLFSSL_CRYPTOCB=yes -DWOLFSSL_CRYPTOCB_RSA_PAD=yes ..
```

Verified locally:

- `--enable-cryptocbutils=rsapad` emits only `-DWOLF_CRYPTO_CB_RSA_PAD`
- `--enable-cryptocbutils` (bare or `=yes`) emits all five (`COPY`, `FREE`, `SETKEY`, `EXPORT_KEY`, `RSA_PAD`)
- `--enable-cryptocbutils=foo` → `Unknown cryptocbutils option: foo. Valid options: copy, free, setkey, export, rsapad`
- `--enable-cryptocbutils=rsapad` without `--enable-cryptocb` → `--enable-cryptocbutils requires --enable-cryptocb`

I also traced the C code end-to-end (rsa.c → cryptocb.c → wc_pkcs11.c) and confirmed the flag actually delivers what the RFE asks for: with it off, PKCS#11 only ever sees `CKM_RSA_X_509`; with it on, the unpadded message + padding metadata crosses the cryptocb boundary and PKCS#11 uses `CKM_RSA_PKCS` / `CKM_RSA_PKCS_OAEP` / `CKM_RSA_PKCS_PSS` with the right `CK_RSA_PKCS_*_PARAMS`. Full trace in the writeup below.

### Patch

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 663c3cf2c..bf93605be 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2208,6 +2208,10 @@ add_option("WOLFSSL_CRYPTOCB_NO_SW_TEST"
     "Disable crypto callback SW testing (default: disabled)"
     "no" "yes;no")
 
+add_option("WOLFSSL_CRYPTOCB_RSA_PAD"
+    "Enable RSA padding crypto callbacks, e.g. for PKCS#11 HSMs that disable raw RSA (default: disabled). Requires WOLFSSL_CRYPTOCB."
+    "no" "yes;no")
+
 add_option("WOLFSSL_PKCALLBACKS"
     "Enable public key callbacks (default: disabled)"
     "no" "yes;no")
@@ -2440,6 +2444,13 @@ if(WOLFSSL_CRYPTOCB_NO_SW_TEST)
     list(APPEND WOLFSSL_DEFINITIONS "-DWC_TEST_NO_CRYPTOCB_SW_TEST")
 endif()
 
+if(WOLFSSL_CRYPTOCB_RSA_PAD)
+    if(NOT WOLFSSL_CRYPTOCB)
+        message(FATAL_ERROR "WOLFSSL_CRYPTOCB_RSA_PAD requires WOLFSSL_CRYPTOCB")
+    endif()
+    list(APPEND WOLFSSL_DEFINITIONS "-DWOLF_CRYPTO_CB_RSA_PAD")
+endif()
+
 # Public Key Callbacks
 if(WOLFSSL_PKCALLBACKS)
     list(APPEND WOLFSSL_DEFINITIONS "-DHAVE_PK_CALLBACKS")
diff --git a/configure.ac b/configure.ac
index 69f8501f0..e9a55202f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -10015,7 +10016,7 @@ fi
 
 # Crypto Callbacks Utils (Copy/Free/etc)
 AC_ARG_ENABLE([cryptocbutils],
-    [AS_HELP_STRING([--enable-cryptocbutils@<:@=copy,free,setkey,export,...@:>@],
+    [AS_HELP_STRING([--enable-cryptocbutils@<:@=copy,free,setkey,export,rsapad,...@:>@],
                     [Enable crypto callback utilities (default: all)])],
     [ ENABLED_CRYPTOCB_UTILS=$enableval ],
     [ ENABLED_CRYPTOCB_UTILS=no ]
@@ -10028,7 +10029,7 @@ if test "$ENABLED_CRYPTOCB_UTILS" != "no"; then
 
     if test "$ENABLED_CRYPTOCB_UTILS" = "yes"; then
         # Enable all utilities
-        AM_CFLAGS="$AM_CFLAGS -DWOLF_CRYPTO_CB_COPY -DWOLF_CRYPTO_CB_FREE -DWOLF_CRYPTO_CB_SETKEY -DWOLF_CRYPTO_CB_EXPORT_KEY"
+        AM_CFLAGS="$AM_CFLAGS -DWOLF_CRYPTO_CB_COPY -DWOLF_CRYPTO_CB_FREE -DWOLF_CRYPTO_CB_SETKEY -DWOLF_CRYPTO_CB_EXPORT_KEY -DWOLF_CRYPTO_CB_RSA_PAD"
     else
         # Parse comma-separated list
         OIFS="$IFS"
@@ -10047,8 +10048,11 @@ if test "$ENABLED_CRYPTOCB_UTILS" != "no"; then
                 export)
                     AM_CFLAGS="$AM_CFLAGS -DWOLF_CRYPTO_CB_EXPORT_KEY"
                     ;;
+                rsapad)
+                    AM_CFLAGS="$AM_CFLAGS -DWOLF_CRYPTO_CB_RSA_PAD"
+                    ;;
                 *)
-                    AC_MSG_ERROR([Unknown cryptocbutils option: $util. Valid options: copy, free, setkey, export])
+                    AC_MSG_ERROR([Unknown cryptocbutils option: $util. Valid options: copy, free, setkey, export, rsapad])
                     ;;
             esac
         done
```

### Full writeup

<details>
<summary>README-issue-10271.md</summary>

````markdown
# Issue #10271 — Expose `WOLF_CRYPTO_CB_RSA_PAD` as a configure knob

## The RFE

Upstream request: <https://github.com/wolfSSL/wolfssl/issues/10271>

`WOLF_CRYPTO_CB_RSA_PAD` gates the RSA padding path in crypto callbacks. When
defined, `wc_pkcs11.c` uses PKCS#11 mechanisms for OAEP / PKCS#1 v1.5 / PSS
directly on the HSM instead of doing raw RSA and padding in software. Many
HSMs disable raw RSA operations, so this flag is effectively required for them
to work well.

Before this change, the macro was only defined implicitly under
`WOLFSSL_RENESAS_TSIP` in `wolfssl/wolfcrypt/settings.h`. Everyone else had to
pass `EXTRA_CFLAGS=-DWOLF_CRYPTO_CB_RSA_PAD`, which is awkward for
distributions (Debian, etc.) and non-obvious for end users.

## Analysis

The codebase already has a good pattern for sub-flags of `--enable-cryptocb`:
`--enable-cryptocbutils=copy,free,setkey,export` — a single option with a
comma-separated list, gated on `--enable-cryptocb` being enabled. Adding a
separate top-level `--enable-crypto-cb-rsa-pad` option would create a second
convention; extending the existing list keeps things consistent.

On the CMake side there is no equivalent composite option — each feature gets
its own `add_option(...)` entry. So the CMake change follows that pattern with
`WOLFSSL_CRYPTOCB_RSA_PAD`.

## Changes

### `configure.ac`

Extended the existing `--enable-cryptocbutils` handling to accept `rsapad`:

- Help string now lists `rsapad`
- `--enable-cryptocbutils=yes` (and bare `--enable-cryptocbutils`) now also
  emits `-DWOLF_CRYPTO_CB_RSA_PAD` alongside the existing four utility defines
- `--enable-cryptocbutils=rsapad` (alone or combined, e.g. `=copy,rsapad`)
  emits just the padding define
- Unknown values are rejected with an updated error message
- The existing precondition `--enable-cryptocbutils requires --enable-cryptocb`
  applies unchanged

### `CMakeLists.txt`

New boolean option `WOLFSSL_CRYPTOCB_RSA_PAD` (default: `no`). When enabled it
emits `-DWOLF_CRYPTO_CB_RSA_PAD`. If set without `WOLFSSL_CRYPTOCB` it fails
at configure time with a clear error message.

## Usage

### Autotools

```sh
# Just the RSA padding callbacks
./configure --enable-cryptocb --enable-cryptocbutils=rsapad

# Combined with other utilities
./configure --enable-cryptocb --enable-cryptocbutils=copy,free,rsapad

# All utilities (now includes rsapad)
./configure --enable-cryptocb --enable-cryptocbutils
```

Typical HSM / PKCS#11 build:

```sh
./configure --enable-pkcs11 --enable-cryptocbutils=rsapad
```

(`--enable-pkcs11` forces `--enable-cryptocb` on, so the precondition is
satisfied automatically.)

### CMake

```sh
cmake -DWOLFSSL_CRYPTOCB=yes -DWOLFSSL_CRYPTOCB_RSA_PAD=yes ..
```

## Examples

Three end-to-end scenarios showing what the flag unlocks in practice.

### 1. Debian packaging

The RFE's motivating case. Debian currently ships without
`WOLF_CRYPTO_CB_RSA_PAD`, so `wc_pkcs11.c` falls back to raw RSA on HSMs that
have it disabled. With this change, the package rules can add a single
configure argument:

```sh
# debian/rules fragment
override_dh_auto_configure:
	dh_auto_configure -- \
	    --enable-pkcs11 \
	    --enable-cryptocbutils=rsapad
```

No more `EXTRA_CFLAGS=-DWOLF_CRYPTO_CB_RSA_PAD` hack in the package rules.

### 2. Application talking to a PKCS#11 HSM

When wolfSSL is routed through a PKCS#11 HSM (SoftHSM for dev, YubiHSM /
Nitrokey / AWS CloudHSM / SafeNet Luna for production), the HSM typically
refuses raw RSA. Enabling `rsapad` makes `wc_pkcs11.c` invoke
`CKM_RSA_PKCS_OAEP`, `CKM_RSA_PKCS`, and `CKM_RSA_PKCS_PSS` mechanisms
directly on the device.

```sh
./configure \
    --enable-pkcs11 \
    --enable-rsapss \
    --enable-cryptocbutils=rsapad
make
```

Application code is unchanged — `wolfSSL_CTX_use_PrivateKey_*` and
`wc_RsaSSL_Sign` / `wc_RsaPSS_Sign` transparently dispatch to the HSM.

### 3. Cross-build via CMake for an embedded target

A typical embedded / yocto-style build using the CMake knob:

```sh
cmake -S . -B build \
    -DCMAKE_TOOLCHAIN_FILE=/opt/toolchains/aarch64.cmake \
    -DWOLFSSL_CRYPTOCB=yes \
    -DWOLFSSL_CRYPTOCB_RSA_PAD=yes \
    -DWOLFSSL_PKCS11=yes
cmake --build build
```

Misconfigurations are caught at configure time. For example, forgetting the
prerequisite:

```sh
$ cmake -S . -B build -DWOLFSSL_CRYPTOCB_RSA_PAD=yes
...
CMake Error: WOLFSSL_CRYPTOCB_RSA_PAD requires WOLFSSL_CRYPTOCB
```

## Verification

Confirmed on a local build:

- `--enable-cryptocbutils=rsapad` emits only `-DWOLF_CRYPTO_CB_RSA_PAD`
- `--enable-cryptocbutils` emits all five (`COPY`, `FREE`, `SETKEY`,
  `EXPORT_KEY`, `RSA_PAD`)
- `--enable-cryptocbutils=foo` fails with
  `Unknown cryptocbutils option: foo. Valid options: copy, free, setkey, export, rsapad`
- `--enable-cryptocbutils=rsapad` without `--enable-cryptocb` fails with
  `--enable-cryptocbutils requires --enable-cryptocb`

## Files touched

- `configure.ac` — `--enable-cryptocbutils` block (help string, all-utils
  define list, case branch, error message)
- `CMakeLists.txt` — new `add_option("WOLFSSL_CRYPTOCB_RSA_PAD" ...)` and
  matching definition emission with a precondition check

## Appendix — C code verification

Traced the runtime effect of the flag end-to-end through the tree to confirm
the configure knob actually delivers what the RFE asks for.

### 1. Caller side — `wolfcrypt/src/rsa.c`

The padded RSA entry points (`wc_RsaPublicEncryptEx` at ~3430,
private-decrypt path at ~3618) gate on `WOLF_CRYPTO_CB_RSA_PAD`:

```c
/* rsa.c:3523 */
if (key->devId != INVALID_DEVID) {
    XMEMSET(&padding, 0, sizeof(RsaPadding));
    padding.pad_value = pad_value;
    padding.pad_type  = pad_type;   /* WC_RSA_PKCSV15_PAD / WC_RSA_PSS_PAD / WC_RSA_OAEP_PAD */
    padding.hash      = hash;
    padding.mgf       = mgf;
    padding.label     = label;
    padding.saltLen   = saltLen;
    ret = wc_CryptoCb_RsaPad(in, inLen, out, &outLen,
                             rsa_type, key, rng, &padding);
```

The `in` buffer here is the **unpadded** message — dispatch happens before the
software padding step, so the plaintext (plus padding metadata) crosses the
cryptocb boundary and the HSM constructs the padded block internally.

The decrypt gate at `rsa.c:3731` routes everything except `RSA_PUBLIC_DECRYPT`
(verify) through the callback. That's correct: HSMs disable raw *private-key*
ops, but public-key verify is cheap and stays in software.

When the macro is off, the old `wc_CryptoCb_Rsa` (no padding argument) is
called instead, and only already-padded bytes reach the callback.

### 2. Dispatcher — `wolfcrypt/src/cryptocb.c:543`

```c
switch (padding->pad_type) {
case WC_RSA_PKCSV15_PAD: pk_type = WC_PK_TYPE_RSA_PKCS; break;
case WC_RSA_PSS_PAD:     pk_type = WC_PK_TYPE_RSA_PSS;  break;
case WC_RSA_OAEP_PAD:    pk_type = WC_PK_TYPE_RSA_OAEP; break;
default:                 pk_type = WC_PK_TYPE_RSA;
}
...
cryptoInfo.pk.rsa.padding = padding;   /* extra struct field,
                                          gated at cryptocb.h:172 */
ret = dev->cb(dev->devId, &cryptoInfo, dev->ctx);
```

Note: `wc_CryptoCb_RsaPad` dispatches to **any** registered crypto callback,
not just PKCS#11. Any backend can choose to handle the new `WC_PK_TYPE_RSA_*`
types; PKCS#11 is just the first consumer.

### 3. PKCS#11 handler — `wolfcrypt/src/wc_pkcs11.c`

Top-level switch at `wc_pkcs11.c:6334` accepts the three new pk types under
the gate and routes to `Pkcs11Rsa` → `Pkcs11RsaEncrypt` / `Pkcs11RsaDecrypt`
/ `Pkcs11RsaSign`. Each maps the pk type to a PKCS#11 mechanism:

| wolfSSL pk_type | PKCS#11 mechanism | Params struct |
|---|---|---|
| `WC_PK_TYPE_RSA_PKCS` | `CKM_RSA_PKCS` | none |
| `WC_PK_TYPE_RSA_OAEP` | `CKM_RSA_PKCS_OAEP` | `CK_RSA_PKCS_OAEP_PARAMS` (hash + mgf from `padding->hash`/`padding->mgf`) |
| `WC_PK_TYPE_RSA_PSS`  | `CKM_RSA_PKCS_PSS`  | `CK_RSA_PKCS_PSS_PARAMS` (hash + mgf + saltLen) |
| `WC_PK_TYPE_RSA` (fallback when macro off) | `CKM_RSA_X_509` | none — **raw RSA** |

Verified at `wc_pkcs11.c:2733-2738` (encrypt), `2815-2820` (decrypt),
`2901-2906` (sign). OAEP / PSS parameter population at `2752-2759`,
`2834-2841`, `2919-2957`.

### Conclusion

- With `WOLF_CRYPTO_CB_RSA_PAD` **off**, a PKCS#11 HSM only ever sees
  `CKM_RSA_X_509` (raw RSA). HSMs that refuse raw RSA will fail.
- With it **on**, the unpadded message + padding metadata crosses the cryptocb
  boundary and the HSM performs OAEP / PKCS#1 v1.5 / PSS internally via the
  matching `CKM_*` mechanism.
- Verify (`RSA_PUBLIC_DECRYPT`) intentionally stays in software even with the
  flag on — public-key ops are cheap and verify doesn't need the HSM.
- Application code is unchanged — dispatch happens inside the padded RSA
  entry points, not at the application layer.

No dead gates, no mismatch between the RFE's stated intent and the actual
implementation. The configure knob is a pure ergonomics improvement: it
exposes correct, already-wired behavior that previously required
`EXTRA_CFLAGS`.
````

</details>

Happy to turn this into a PR if useful.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2243,6 +2243,10 @@ add_option("WOLFSSL_CRYPTOCB_NO_SW_TEST"
     "Disable crypto callback SW testing (default: disabled)"
     "no" "yes;no")
 
+add_option("WOLFSSL_CRYPTOCB_RSA_PAD"
+    "Enable RSA padding aware crypto callbacks (default: disabled)"
+    "no" "yes;no")
+
 add_option("WOLFSSL_PKCALLBACKS"
     "Enable public key callbacks (default: disabled)"
     "no" "yes;no")
@@ -2473,6 +2477,13 @@ endif()
 
 if(WOLFSSL_CRYPTOCB_NO_SW_TEST)
     list(APPEND WOLFSSL_DEFINITIONS "-DWC_TEST_NO_CRYPTOCB_SW_TEST")
+endif()
+
+if(WOLFSSL_CRYPTOCB_RSA_PAD)
+    if(NOT WOLFSSL_CRYPTOCB)
+        message(FATAL_ERROR "WOLFSSL_CRYPTOCB_RSA_PAD requires WOLFSSL_CRYPTOCB")
+    endif()
+    list(APPEND WOLFSSL_DEFINITIONS "-DWOLF_CRYPTO_CB_RSA_PAD")
 endif()
 
 # Public Key Callbacks

--- a/cmake/options.h.in
+++ b/cmake/options.h.in
@@ -258,6 +258,8 @@ extern "C" {
 #cmakedefine WC_RSA_PSS
 #undef WOLF_CRYPTO_CB
 #cmakedefine WOLF_CRYPTO_CB
+#undef WOLF_CRYPTO_CB_RSA_PAD
+#cmakedefine WOLF_CRYPTO_CB_RSA_PAD
 #undef WOLFSSL_AARCH64_BUILD
 #cmakedefine WOLFSSL_AARCH64_BUILD
 #undef WOLFSSL_AES_CFB

--- a/configure.ac
+++ b/configure.ac
@@ -10647,6 +10647,24 @@ if test "$ENABLED_CRYPTOCB_UTILS" != "no"; then
     fi
 fi
 
+# Crypto callback RSA padding support
+# When enabled, the RSA crypto callback args struct exposes the RsaPadding
+# parameters so the callback can perform RSA padding/unpadding itself or
+# offload it together with the modular exponentiation.
+AC_ARG_ENABLE([cryptocb-rsa-pad],
+    [AS_HELP_STRING([--enable-cryptocb-rsa-pad],[Enable RSA padding aware crypto callbacks (default: disabled). Requires --enable-cryptocb])],
+    [ ENABLED_CRYPTOCB_RSA_PAD=$enableval ],
+    [ ENABLED_CRYPTOCB_RSA_PAD=no ]
+    )
+
+if test "$ENABLED_CRYPTOCB_RSA_PAD" = "yes"
+then
+    if test "x$ENABLED_CRYPTOCB" = "xno"; then
+        AC_MSG_ERROR([--enable-cryptocb-rsa-pad requires --enable-cryptocb])
+    fi
+    AM_CFLAGS="$AM_CFLAGS -DWOLF_CRYPTO_CB_RSA_PAD"
+fi
+
 
 # Asynchronous Crypto
 AC_ARG_ENABLE([asynccrypt],

--- a/src/tls13.c
+++ b/src/tls13.c
@@ -15147,7 +15147,15 @@ const char* wolfSSL_get_cipher_name_by_hash(WOLFSSL* ssl, const char* hash)
     const char* name = NULL;
     byte mac = no_mac;
     int i;
-    const Suites* suites = WOLFSSL_SUITES(ssl);
+    const Suites* suites;
+
+    if (hash == NULL || ssl == NULL ||
+        (ssl->suites == NULL && ssl->ctx == NULL))
+        return NULL;
+
+    suites = WOLFSSL_SUITES(ssl);
+    if (suites == NULL)
+        return NULL;
 
     if (XSTRCMP(hash, "SHA256") == 0) {
         mac = sha256_mac;

--- a/src/tls13.c
+++ b/src/tls13.c
@@ -995,6 +995,18 @@ int Tls13_Exporter(WOLFSSL* ssl, unsigned char *out, size_t outLen,
     }
 #endif /* WOLFSSL_DTLS13 */
 
+    /* Sanity check contextLen to prevent truncation when cast to word32. */
+    if (contextLen > WOLFSSL_MAX_32BIT)
+        return BAD_FUNC_ARG;
+    /* RFC 8446 HkdfLabel encodes the output length as a uint16, so requested
+     * lengths > 65535 cannot be represented and must be rejected. */
+    if (outLen > WOLFSSL_MAX_16BIT)
+        return BAD_FUNC_ARG;
+    /* RFC 8446 HkdfLabel encodes the label length in a single byte, so
+     * anything > 255 cannot be represented and must be rejected. */
+    if (labelLen > (WOLFSSL_MAX_8BIT - protocolLen))
+        return BAD_FUNC_ARG;
+
     switch (ssl->specs.mac_algorithm) {
         #ifndef NO_SHA256
         case sha256_mac:
@@ -1039,11 +1051,6 @@ int Tls13_Exporter(WOLFSSL* ssl, unsigned char *out, size_t outLen,
             emptyHash, hashLen, (int)hashType);
     if (ret != 0)
         return ret;
-
-    /* Sanity check contextLen to prevent truncation when cast to word32. */
-    if (contextLen > WOLFSSL_MAX_32BIT) {
-        return BAD_FUNC_ARG;
-    }
 
     /* Hash(context_value) */
     ret = wc_Hash(hashType, context, (word32)contextLen, hashOut, WC_MAX_DIGEST_SIZE);

--- a/src/tls13.c
+++ b/src/tls13.c
@@ -6053,7 +6053,8 @@ static int DoTls13CertificateRequest(WOLFSSL* ssl, const byte* input,
     int         ret = 0;
     Suites      peerSuites;
 #ifdef WOLFSSL_POST_HANDSHAKE_AUTH
-    CertReqCtx* certReqCtx;
+    word16      reqCtxLen;
+    const byte* reqCtxData;
 #endif
 
     WOLFSSL_START(WC_FUNC_CERTIFICATE_REQUEST_DO);
@@ -6077,18 +6078,12 @@ static int DoTls13CertificateRequest(WOLFSSL* ssl, const byte* input,
         return BUFFER_ERROR;
 
 #ifdef WOLFSSL_POST_HANDSHAKE_AUTH
-    /* CertReqCtx has one byte at end for context value.
-     * Increase size to handle other implementations sending more than one byte.
-     * That is, allocate extra space, over one byte, to hold the context value.
+    /* Remember the request context bytes; the CertReqCtx allocation and
+     * linking into ssl->certReqCtx is deferred until after the rest of the
+     * message has been validated.
      */
-    certReqCtx = (CertReqCtx*)XMALLOC(sizeof(CertReqCtx) + (len == 0 ? 0 : len - 1), ssl->heap,
-                                                       DYNAMIC_TYPE_TMP_BUFFER);
-    if (certReqCtx == NULL)
-        return MEMORY_E;
-    certReqCtx->next = ssl->certReqCtx;
-    certReqCtx->len = len;
-    XMEMCPY(&certReqCtx->ctx, input + *inOutIdx, len);
-    ssl->certReqCtx = certReqCtx;
+    reqCtxLen = len;
+    reqCtxData = input + *inOutIdx;
 #endif
     *inOutIdx += len;
 
@@ -6154,6 +6149,24 @@ static int DoTls13CertificateRequest(WOLFSSL* ssl, const byte* input,
 
     /* This message is always encrypted so add encryption padding. */
     *inOutIdx += ssl->keys.padSz;
+
+#ifdef WOLFSSL_POST_HANDSHAKE_AUTH
+    {
+    /* CertReqCtx has one byte at end for context value.
+     * Increase size to handle other implementations sending more than one byte.
+     * That is, allocate extra space, over one byte, to hold the context value.
+     */
+        CertReqCtx* certReqCtx = (CertReqCtx*)XMALLOC(
+            sizeof(CertReqCtx) + (reqCtxLen == 0 ? 0 : reqCtxLen - 1),
+            ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        if (certReqCtx == NULL)
+            return MEMORY_E;
+        certReqCtx->next = ssl->certReqCtx;
+        certReqCtx->len = reqCtxLen;
+        XMEMCPY(&certReqCtx->ctx, reqCtxData, reqCtxLen);
+        ssl->certReqCtx = certReqCtx;
+    }
+#endif
 
     WOLFSSL_LEAVE("DoTls13CertificateRequest", ret);
     WOLFSSL_END(WC_FUNC_CERTIFICATE_REQUEST_DO);

--- a/src/tls13.c
+++ b/src/tls13.c
@@ -1003,8 +1003,9 @@ int Tls13_Exporter(WOLFSSL* ssl, unsigned char *out, size_t outLen,
     if (outLen > WOLFSSL_MAX_16BIT)
         return BAD_FUNC_ARG;
     /* RFC 8446 HkdfLabel encodes the label length in a single byte, so
-     * anything > 255 cannot be represented and must be rejected. */
-    if (labelLen > (WOLFSSL_MAX_8BIT - protocolLen))
+     * anything > 255 cannot be represented and must be rejected.
+     * The protocol length is included in the label. */
+    if ((labelLen +  protocolLen) > WOLFSSL_MAX_8BIT)
         return BAD_FUNC_ARG;
 
     switch (ssl->specs.mac_algorithm) {

--- a/src/tls13.c
+++ b/src/tls13.c
@@ -6153,10 +6153,10 @@ static int DoTls13CertificateRequest(WOLFSSL* ssl, const byte* input,
 
 #ifdef WOLFSSL_POST_HANDSHAKE_AUTH
     {
-    /* CertReqCtx has one byte at end for context value.
-     * Increase size to handle other implementations sending more than one byte.
-     * That is, allocate extra space, over one byte, to hold the context value.
-     */
+        /* CertReqCtx has one byte at end for context value.
+        * Increase size to handle other implementations sending more than one byte.
+        * That is, allocate extra space, over one byte, to hold the context value.
+        */
         CertReqCtx* certReqCtx = (CertReqCtx*)XMALLOC(
             sizeof(CertReqCtx) + (reqCtxLen == 0 ? 0 : reqCtxLen - 1),
             ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);

